File: LeetCode/Tree/TraverseTree/Leetcode_112_path_sum/Leetcode_112_path_sum.cpp
/**
 * @brief: Leetcode_112_路径总和
 * @link: https://leetcode.cn/problems/path-sum/description/
 * @author: liangj.zhang
 * @date: 6/11/2025
 * 
 * @Difficulty: Easy
 * 
 * @Label: Tree(Traverse)
 * 
 * @Retrospect(worthy 1 - 5): 3
 * 
 * @thoughts:
 *  + 【思路 1】：递归
 *      + 【写法 1】：在叶子结点结算
 *          + 分析：
 *              + 时间复杂度：O(n)
 *              + 空间复杂度：O(h)，h 为树的高度
 *          + rank:
 *              + 时间效率：0 ms, 击败 100%
 *              + 空间效率：20.96 MB, 击败 92.86%
 * 
 *      + 【写法 2】：在空节点结算，但需要额外借助一个 parent 变量；
 *          + 分析：
 *              + 时间复杂度：O(n)
 *              + 空间复杂度：O(h)，h 为树的高度
 *          + rank:
 *              + 时间效率：0 ms, 击败 100%
 *              + 空间效率：21.05 MB, 击败 65.06%
 */

#include "../../../../Include/Leetcode/Tree/Tree.h"
using Leetcode::Tree::BinaryTree::TreeNode;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// 【写法 2】：也是递归，在空节点结算；
class Solution {
private:
    bool find;

    void __hasPathSum(TreeNode* root, int left, TreeNode* parent) {

        if (!root) {

            if (!find && left == 0 && !parent->left && !parent->right)
                find = true;

            return;
        }

        if (!find) {
            
            __hasPathSum(root->left, left - root->val, root);
            __hasPathSum(root->right, left - root->val, root);
        }
    }

public:
    bool hasPathSum(TreeNode* root, int targetSum) {

        if (!root) return false;
        
        find = false;
        __hasPathSum(root, targetSum, NULL);

        return find;
    }
};

 // 【写法 1】：递归，在叶子结点结算
class Solution {
private:
    bool find;

    void __hasPathSum(TreeNode* root, int left) {

        if (!root->left && !root->right && left == 0) 
            find = true;

        if (!find) {
            
            if (root->left) __hasPathSum(root->left, left - root->left->val);
            if (root->right) __hasPathSum(root->right, left - root->right->val);
        }
    }

public:
    bool hasPathSum(TreeNode* root, int targetSum) {

        if (!root) return false;
        
        find = false;
        __hasPathSum(root, targetSum - root->val);

        return find;
    }
};

File: LeetCode_Python/Tree/TraverseTree/Leetcode_112_path_sum/Leetcode_112_path_sum.py
"""
@brief: Leetcode_112_路径总和  
@link: https://leetcode.cn/problems/path-sum/description/  
@author: liangj.zhang  
@date: 6/11/2025  

@Difficulty: Easy  

@Label: Tree(Traverse)  

@Retrospect(worthy 1 - 5): 3  

@thoughts:
 + 【思路 1】：递归  
     + 【写法 1】：在叶子节点结算  
         + 分析：  
             + 时间复杂度：O(n)  
             + 空间复杂度：O(h)，h 为树的高度  
         + rank:  
             + 时间效率：0 ms, 击败 100%  
             + 空间效率：18.71 MB, 击败 36.78%
"""

import os
import sys

print(os.path.join(os.path.dirname(__file__)))
print(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../.."))) #f:\Code\zljgit\LeetCode
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../..")))

from typing import Optional

from Include.PyLeetcode.Tree.Tree import TreeNode

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def __init__(self):
        self.find = False

    def __hasPathSum(self, root: Optional[TreeNode], left: int) -> None:
        
        if not root.left and not root.right and left == 0:
            self.find = True
        
        if not self.find:
            if root.left: self.__hasPathSum(root.left, left - root.left.val)
            if root.right: self.__hasPathSum(root.right, left - root.right.val)

    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root: return False
        self.find = False
        self.__hasPathSum(root, targetSum - root.val)
        return self.find